### PR TITLE
MYAC-357 Reserve IP

### PR DESCRIPTION
**Pull request by nlgordon**

### CODE REVIEW

File: engine/schema/resources/META-INF/cloudstack/core/spring-engine-schema-core-daos-context.xml [hunk 1]
```diff
@@ -220,6 +220,7 @@
   <bean id="nicSecondaryIpDaoImpl" class="com.cloud.vm.dao.NicSecondaryIpDaoImpl" />
   <bean id="nicIpAliasDaoImpl" class="com.cloud.vm.dao.NicIpAliasDaoImpl" />
   <bean id="objectInDataStoreDaoImpl" class="org.apache.cloudstack.storage.db.ObjectInDataStoreDaoImpl" />
+  <bean id="ontapSvmDaoImpl" class="com.cloud.network.dao.IpReservationDaoImpl" />
```
Comment (Slair1) [inline engine/schema/resources/META-INF/cloudstack/core/spring-engine-schema-core-daos-context.xml]: Can we leave the IP Reservation functionality generic?  (Not mention ontap)

File: engine/schema/src/com/cloud/network/IpReservationVO.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        return removed;
+    }
+
+    @Override
+    public void setRemoved(Date removed) {
+        this.removed = removed;
+    }
+
+    public String toString() {
+        return "OntapSvm[" + id +
```
Comment (Slair1) [inline engine/schema/src/com/cloud/network/IpReservationVO.java]: Same here, can we make the IP Reservation functionality generic and not mention ONTAP?

File: server/src/com/cloud/network/NetworkModelImpl.java [hunk 3]
```diff
@@ -1872,6 +1876,11 @@ public List<String> getUsedIpsInNetwork(Network network) {
         //Get ips used by load balancers
         List<String> lbIps = _appLbRuleDao.listLbIpsBySourceIpNetworkId(network.getId());
         ips.addAll(lbIps);
+        // Get ips used by Ontap SVMs
```
Comment (Slair1) [inline server/src/com/cloud/network/NetworkModelImpl.java]: Same here, make generic, remove reference to ONTAP

**Comment by Slair1** [comment]
@nlgordon thanks Nate!  I made a few minor comments during my cursory review.  Let's make sure it's all generic and we don't mention ONTAP.  in the future anything could use this feature.

Also, how much permission checking is done to ensure the network ID specified is being modified by someone with the right permissions in the domain/account hierarchy?

**Comment by nlgordon** [comment]
@Slair1 Right now I've limited the API to root admins only. Since this will only be called from MYAC backend for the moment, we don't need more complex permissions. That can always be added later when we add UI for this, but for right now it makes it far simpler.

File: engine/schema/resources/META-INF/cloudstack/core/spring-engine-schema-core-daos-context.xml (hunk 1 above)
Comment (nlgordon) [reply engine/schema/resources/META-INF/cloudstack/core/spring-engine-schema-core-daos-context.xml]: Done

File: engine/schema/src/com/cloud/network/IpReservationVO.java (hunk 2 above)
Comment (nlgordon) [reply engine/schema/src/com/cloud/network/IpReservationVO.java]: Done

File: server/src/com/cloud/network/NetworkModelImpl.java (hunk 3 above)
Comment (nlgordon) [reply server/src/com/cloud/network/NetworkModelImpl.java]: Done

**Comment by Slair1** [comment]
@nlgordon During or meeting we said to update this so we could reserve a specific IP or ask for an IP and have it reserved, correct?

**Comment by nlgordon** [comment]
@Slair1 Yes, we did say we wanted to be able to reserve a specific IP. After looking at it again, it already has the ability to reserve a range of specific IPs. Technically all reservations are for a range. We also talked about having this be a user facing functionality as well. Do you want to include the ACS UI and user permissions work in this, or break that out as a separate task/branch?

**Comment by Slair1** [comment]
> @Slair1 Yes, we did say we wanted to be able to reserve a specific IP. After looking at it again, it already has the ability to reserve a range of specific IPs. Technically all reservations are for a range. We also talked about having this be a user facing functionality as well. Do you want to include the ACS UI and user permissions work in this, or break that out as a separate task/branch?

@nlgordon I think we include the ACS UI and user permission into a future task/branch.  Do you think this is ready for merge then as is?

**Comment by nlgordon** [comment]
@Slair1 I believe we are good to merge this now. We also have the MYAC part ready to go as well.

**Review by Slair1 — APPROVED** [review]

**Pull request merged**